zkevm-circuits/src/evm_circuit/execution/calldataload.rs

```diff
@@ -0,0 +1,312 @@
+use halo2::{arithmetic::FieldExt, plonk::Error};
+
+use crate::{
+    evm_circuit::{
+        param::{N_BYTES_MEMORY_ADDRESS, N_BYTES_WORD},
+        step::ExecutionState,
+        table::{CallContextFieldTag, TxContextFieldTag},
+        util::{
+            common_gadget::SameContextGadget,
+            constraint_builder::{ConstraintBuilder, StepStateTransition, Transition},
+            from_bytes,
+            memory_gadget::BufferReaderGadget,
+            Cell, RandomLinearCombination,
+        },
+        witness::{Block, Call, ExecStep, Transaction},
+    },
+    util::Expr,
+};
+
+use super::ExecutionGadget;
+
+#[derive(Clone, Debug)]
+pub(crate) struct CallDataLoadGadget<F> {
+    /// Gadget to constrain the same context.
+    same_context: SameContextGadget<F>,
+    /// Transaction id from the tx context.
+    tx_id: Cell<F>,
+    /// The bytes offset in calldata, from which we load a 32-bytes word.
+    calldata_start: Cell<F>,
+    /// The bytes offset in calldata, where we end the 32-bytes word.
+    calldata_end: Cell<F>,
+    /// Gadget to read from tx calldata, which we validate against the word
+    /// pushed to stack.
+    buffer_reader: BufferReaderGadget<F, N_BYTES_WORD, N_BYTES_MEMORY_ADDRESS>,
+}
+
+impl<F: FieldExt> ExecutionGadget<F> for CallDataLoadGadget<F> {
+    const EXECUTION_STATE: ExecutionState = ExecutionState::CALLDATALOAD;
+
+    const NAME: &'static str = "CALLDATALOAD";
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+
+        let calldata_start = cb.query_cell();
+        let calldata_end = cb.query_cell();
+
+        // Pop the offset value from stack.
+        cb.stack_pop(from_bytes::expr(&[calldata_start.clone()]));
+
+        // Add a lookup constrain for TxId in the RW table.
+        let tx_id = cb.call_context(None, CallContextFieldTag::TxId);
+
+        let buffer_reader = BufferReaderGadget::construct(cb, &calldata_start, &calldata_end);
+
+        // Initialise all data bytes to 0.
+        let mut calldata_word = [0u8; N_BYTES_WORD].map(|i| i.expr());
+        for (i, data) in calldata_word.iter_mut().enumerate() {
+            // If the buffer is readable, add a lookup constraint for that next
```

Comment:
you can simplify the logic for this part as follows
```
let mut calldata_word = (0..N_BYTES_WORD).map(|i| {
  let byte = buffer_eader.byte(i);
  // lookup...
  byte
}).collect();
```

Reply:
This is because if the `read_flag` is 0, the `BufferReaderGadget` will constrain the `byte` to be 0.